crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+                .inspect(|path| trace!("Found candidate Python interpreter: {}", path.display()))
+                .chain(
+                    // TODO(zanieb): Consider moving `python.bat` into `possible_names` to avoid a chain
+                    cfg!(windows)
+                        .then(move || {
+                            which::which_in_global("python.bat", Some(&dir_clone))
+                                .into_iter()
+                                .flatten()
+                                .collect::<Vec<_>>()
+                        })
+                        .into_iter()
+                        .flatten(),
+                )
+        })
+}
+
+/// Lazily iterate over all discoverable Python interpreters.
+///
+///See [`python_executables`] for more information on discovery.
+fn python_interpreters<'a>(
+    version: Option<&'a VersionRequest>,
+    sources: &'a SourceSelector,
+    cache: &'a Cache,
+) -> impl Iterator<Item = Result<(InterpreterSource, Interpreter), Error>> + 'a {
+    let iter = python_executables(version, sources).map(|result| match result {
+        Ok((source, path)) => Interpreter::query(path, cache)
+            .map(|interpreter| (source, interpreter))
+            .map_err(Error::from),
+        Err(err) => Err(err),
+    });
+
+    iter
+}
+
+/// Find an interpreter that satisfies the given request.
+///
+/// If an error is encountered while locating or inspecting a candidate interpreter,
+/// the error will raised instead of attempting further candidates.
+pub fn find_interpreter(
+    request: &InterpreterRequest,
+    sources: &SourceSelector,
+    cache: &Cache,
+) -> Result<InterpreterResult, Error> {
+    let result = match request {
+        InterpreterRequest::File(path) => {
+            if !sources.contains(InterpreterSource::ProvidedPath) {
+                return Err(Error::SourceNotSelected(
+                    request.clone(),
+                    InterpreterSource::ProvidedPath,
+                ));
+            }
+            if !path.try_exists()? {
+                return Ok(NoInterpreter(InterpreterNotFound::FileNotFound(
+                    path.clone(),
+                )));
+            }
+            DiscoveredInterpreter {
+                source: InterpreterSource::ProvidedPath,
+                interpreter: Interpreter::query(path, cache)?,
+            }
+        }
+        InterpreterRequest::Directory(path) => {
+            if !sources.contains(InterpreterSource::ProvidedPath) {
+                return Err(Error::SourceNotSelected(
+                    request.clone(),
+                    InterpreterSource::ProvidedPath,
+                ));
+            }
+            if !path.try_exists()? {
+                return Ok(NoInterpreter(InterpreterNotFound::FileNotFound(
+                    path.clone(),
+                )));
+            }
+            let executable = virtualenv_python_executable(path);
+            if !executable.try_exists()? {
```

Comment:
What's the motivation for `try_exists` over `exists`? Just wondering.

Reply:
Just because the Rust docs suggest that `exists` is error prone and `try_exists` is better to use. I think if we encounter a permissions error on a parent directory it'll fail instead of saying `false`.